File: frontend/src/api/languages.ts
import client from "./client";

export interface Language {
  lang_code: string;
  language_name: string;
  script: string;
  region: string;
  continent: string;
  family: string;
  latitude: number;
  longitude: number;
  training_hours: number;
  cer_7b_llm: number;
  cer_7b_ctc: number;
  endangerment: string;
}

export interface LanguageListResponse {
  total: number;
  page: number;
  page_size: number;
  languages: Language[];
}

export interface LanguageDetail extends Language {
  cer_by_model: Record<string, number>;
}

export interface LanguageFilters {
  region?: string;
  script?: string;
  cer_max?: number;
  endangered?: string;
  search?: string;
  sort_by?: string;
  sort_desc?: boolean;
  page?: number;
  page_size?: number;
}

export async function getLanguages(filters: LanguageFilters = {}): Promise<LanguageListResponse> {
  const params = new URLSearchParams();
  if (filters.region) params.set("region", filters.region);
  if (filters.script) params.set("script", filters.script);
  if (filters.cer_max !== undefined) params.set("cer_max", String(filters.cer_max));
  if (filters.endangered) params.set("endangered", filters.endangered);
  if (filters.search) params.set("search", filters.search);
  if (filters.sort_by) params.set("sort_by", filters.sort_by);
  if (filters.sort_desc) params.set("sort_desc", "true");
  if (filters.page) params.set("page", String(filters.page));
  if (filters.page_size) params.set("page_size", String(filters.page_size));

  const { data } = await client.get<LanguageListResponse>(`/languages?${params.toString()}`);
  return data;
}

export async function getLanguage(code: string): Promise<LanguageDetail> {
  const { data } = await client.get<LanguageDetail>(`/languages/${code}`);
  return data;
}

export interface MapPoint {
  lang_code: string;
  language_name: string;
  latitude: number;
  longitude: number;
  cer: number;
  endangerment: string;
  continent: string;
  training_hours: number;
}

export async function getMapPoints(): Promise<MapPoint[]> {
  const { data } = await client.get<MapPoint[]>("/languages/map-points");
  return data;
}

export async function getScripts(): Promise<string[]> {
  const { data } = await client.get<string[]>("/languages/scripts");
  return data;
}

export async function getContinents(): Promise<string[]> {
  const { data } = await client.get<string[]>("/languages/continents");
  return data;
}

File: backend/app/routers/languages.py
from fastapi import APIRouter, Request, Query, HTTPException
from app.models.language import LanguageBase, LanguageListResponse, LanguageDetailResponse

router = APIRouter(tags=["languages"])


@router.get("/languages", response_model=LanguageListResponse)
async def list_languages(
    request: Request,
    region: str | None = None,
    script: str | None = None,
    cer_max: float | None = None,
    endangered: str | None = None,
    search: str | None = None,
    sort_by: str = "lang_code",
    sort_desc: bool = False,
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=500),
):
    data_service = request.app.state.data_service
    languages, total = data_service.get_all_languages(
        region=region,
        script=script,
        cer_max=cer_max,
        endangered=endangered,
        search=search,
        sort_by=sort_by,
        sort_desc=sort_desc,
        page=page,
        page_size=page_size,
    )
    return LanguageListResponse(
        total=total,
        page=page,
        page_size=page_size,
        languages=[LanguageBase(**lang.__dict__) for lang in languages],
    )


@router.get("/languages/map-points")
async def get_map_points(request: Request):
    return request.app.state.data_service.get_map_points()


@router.get("/languages/scripts", response_model=list[str])
async def list_scripts(request: Request):
    return request.app.state.data_service.get_unique_scripts()


@router.get("/languages/continents", response_model=list[str])
async def list_continents(request: Request):
    return request.app.state.data_service.get_unique_continents()


@router.get("/languages/{lang_code}", response_model=LanguageDetailResponse)
async def get_language(request: Request, lang_code: str):
    data_service = request.app.state.data_service
    lang = data_service.get_language(lang_code)
    if not lang:
        raise HTTPException(status_code=404, detail=f"Language '{lang_code}' not found")
    return LanguageDetailResponse(**lang.__dict__)
